### PR TITLE
Data Availability Submission API

### PR DESCRIPTION
**Pull request by tzdybal**
I'm splitting #53 into multiple self-contained PRs. This is the first one.

This PR introduces Universal DA interface, only the submission method. It also starts dummy-mock implementation.

Note: I'll definitely squash before merging ;)

rendered ADR: https://github.com/lazyledger/optimint/blob/18370e0b258fabf846a637b98a20cd6abe4d6e27/docs/lazy-adr/adr-005-da-interface.md

### CODE REVIEW

**Review by liamsi — COMMENTED** [review]
Left some comments.

**Review by Wondertan — APPROVED** [review]

File: docs/lazy-adr/adr-005-da-interface.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+Optimint requires data availability layer. Different implementations are expected.
+
+## Alternative Approaches
+
+> This section contains information around alternative options that are considered before making a decision. It should contain a explanation on why the alternative approach(es) were not chosen.
+
+## Decision
+
+Defined interface should be very generic.
+Interface should consist of 3 methods: `Start`, `Stop`, `SubmitBlock`.
```
[the remark this answers is not in the document]
Comment (tzdybal) [reply docs/lazy-adr/adr-005-da-interface.md:18]: ADR is out-of-date. I'll fix this.

File: da/da.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	// parts of Optimint.
+	// Hash hash.Hash
+}
+
+type DataAvailabilityLayerClient interface {
+	// Init is called once to allow DA client to read configuration and initialize resources.
+	Init(config []byte, logger log.Logger) error
+
+	Start() error
+	Stop() error
```
[the remark this answers is not in the document]
Comment (tzdybal) [reply da/da.go:34]: I think it's a good practice to give implementation a clear life-cycle - especially graceful stopping may be important. Alternatively we may switch to using `Context` for this.

Definitely it should be better documented, and "Howto add your fancy DA layer to optimint" is a great idea. Created #74 to keep this in mind.

File: da/da.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	"github.com/lazyledger/optimint/types"
+)
+
+// TODO define an enum of different non-happy-path cases
+// that might need to be handled by Optimint independent of
+// the underlying DA chain.
+type StatusCode uint64
+
+const (
+	StatusSuccess StatusCode = iota
```
[the remark this answers is not in the document]
Comment (tzdybal) [reply da/da.go:14]: I was not thinking particularly about protobuf, but rather 'UNIX style' (0 means no error). Maybe we should add `StatusUnknown` as 0? What do you think?

[the remark this answers is not in the document]
Comment (liamsi) [reply da/da.go:14]: Yeah, I think in that case that's is fine.

**Pull request merged**